### ISSUE TITLE
JPADatabase generated class with default schema "PUBILIC", not working in mysql.

### ISSUE THREAD

**Issue by scolia** [jOOQ/jOOQ#9013]
### Your question:
I am using jooq-codegen-maven to generate class from jpa entity: 

```
          <plugin>
                <groupId>org.jooq</groupId>
                <artifactId>jooq-codegen-maven</artifactId>
                <version>3.11.11</version>
                <dependencies>
                    <dependency>
                        <groupId>org.jooq</groupId>
                        <artifactId>jooq-meta-extensions</artifactId>
                        <version>3.11.11</version>
                    </dependency>
                </dependencies>
                <executions>
                    <execution>
                        <id>default-cli</id>
                        <phase>generate-sources</phase>
                        <goals>
                            <goal>generate</goal>
                        </goals>
                    </execution>
                </executions>
                <configuration>
                    <generator>
                        <database>
                            <name>org.jooq.meta.extensions.jpa.JPADatabase</name>
                            <properties>
                                <!-- A comma separated list of Java packages, that contain your entities -->
                                <property>
                                    <key>packages</key>
                                    <value>com.test.web.entity</value>
                                </property>
                            </properties>
                        </database>
                    </generator>
                </configuration>
            </plugin>
```

And then i using DSLContext to query something:

```
public User findByUserName(String userName) {
        return dslContext.select()
                .from(USER)
                .where(USER.USERNAME.eq(userName))
                .fetchOneInto(User.class);
}
```

I get error. Can not find the table name: PUBLIC.USER

Because the generated  class with default schema "PUBLIC", but in mysql, i have not schema name "PUBLIC".

how can i fix this problem?

### Versions:

- jOOQ: 3.11.11
- Java: jdk8
- Database (include vendor): mysql
- OS:
- JDBC Driver (include name if inofficial driver):

**Comment by knutwannheden** [comment]
The schema reference can be suppresed in the generated SQL. All you have to do is this, wherever you create your `DSLContext`:

```java
dslContext.configuration().settings().withRenderSchema(false);
```
It would however be even cleaner if you configured this setting _before_ creating the `DSLContext` instance using one of the overloaded `DSL#using()` methods with a `Settings` parameter.

Please reopen if I missed something.

**Comment by scolia** [comment]
Thank you for your help. I also solved this problem by modifying the configuration.

```
 <inputSchema>PUBLIC</inputSchema>
 <outputSchemaToDefault>true</outputSchemaToDefault>
```

But now I face a new problem.

```
bad SQL grammar [select USER.USER_ID, USER.PASSWORD, USER.USER_NAME from USER where USER.USER_NAME = ?];
```
All table name and field is Uppercase. 

This table is created by jpa(hibernate) in spring boot. so the table name is **user**, and field is **user_id**, **user_name**, **password**.

How can i change any setting to solve this problem?

Thanks!

**Comment by knutwannheden** [comment]
Yes, of course using `outputSchemaToDefault` is even better here. I had forgotten about this flag.

For the other problem: That sounds a bit weird. The jOOQ code generator should generate Java code using the same names as passed on by Hibernate (i.e. as in the JPA annotations). Can you check if the Java strings in the generated code look correct?

As a workaround you may want to use `Settings#withRenderNameStyle()`. Try with `AS_IS` or `LOWER`.

**Comment by scolia** [comment]
@knutwannheden 

This is my entity:

```
@Getter
@Setter
@ToString
@Entity
@Table(name = "user")
public class User {

    @Id
    @Column(name = "user_id")
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long userId;

    @Column(name = "user_name", nullable = false)
    private String userName;

    @Column(name = "password", nullable = false)
    private String password;
}
```

This is the part of generate class:

```
  /**
     * The reference instance of <code>USER</code>
     */
    public static final JUser USER = new JUser();

    /**
     * The class holding records for this type
     */
    @Override
    public Class<JUserRecord> getRecordType() {
        return JUserRecord.class;
    }

    /**
     * The column <code>USER.USER_ID</code>.
     */
    public final TableField<JUserRecord, Long> USER_ID = createField("USER_ID", org.jooq.impl.SQLDataType.BIGINT.nullable(false).identity(true), this, "");

    /**
     * The column <code>USER.PASSWORD</code>.
     */
    public final TableField<JUserRecord, String> PASSWORD = createField("PASSWORD", org.jooq.impl.SQLDataType.VARCHAR(255).nullable(false), this, "");

    /**
     * The column <code>USER.USER_NAME</code>.
     */
    public final TableField<JUserRecord, String> USER_NAME = createField("USER_NAME", org.jooq.impl.SQLDataType.VARCHAR(255).nullable(false), this, "");

    /**
     * Create a <code>USER</code> table reference
     */
    public JUser() {
        this(DSL.name("USER"), null);
    }

```

if i change **user_name** to **username**, than the generate class will change to:

```
    /**
     * The column <code>USER.USERNAME</code>.
     */
    public final TableField<JUserRecord, String> USERNAME = createField("USERNAME", org.jooq.impl.SQLDataType.VARCHAR(255).nullable(false), this, "");
```

**Comment by knutwannheden** [comment]
Thanks. This indeed looks like a bug somewhere. I will create a separate issue for this.

Did you try the workaround I proposed?

**Comment by knutwannheden** [comment]
It looks like your problem with the identifiers could be related to #6709. Do you know which "hibernate.physical_naming_strategy" you are using? With jOOQ 3.12 you would be able to configure this in the generator:

```xml
<database>
    <name>org.jooq.meta.extensions.jpa.JPADatabase</name>
    <properties>
        <property>
            <key>packages</key>
            <value>org.jooq.codegen.test.jpa</value>
        </property>
        <property>
            <key>hibernate.physical_naming_strategy</key>
            <value>org.springframework.boot.orm.jpa.hibernate.SpringPhysicalNamingStrategy</value>
        </property>
    </properties>
    <includes>.*</includes>
</database>
```

**Comment by scolia** [comment]
Set **Settings#withRenderNameStyle()** to **LOWER** is work. But config the generator look more graceful. 

Any plans to release 3.12?

**Comment by knutwannheden** [comment]
> Any plans to release 3.12?

We expect to release 3.12 later this month, but no promises.

**Comment by scolia** [comment]
@knutwannheden 

I just clone master to local, install in my local maven repository. After some test. I think i find some bug.

first, my maven plugin config look like:

```
  <property>
    <key>hibernate.physical_naming_strategy</key> 
    <value>org.springframework.boot.orm.jpa.hibernate.SpringPhysicalNamingStrategy</value>
  </property>
  <property>
     <key>hibernate.implicit_naming_strategy</key> 
     <value>org.springframework.boot.orm.jpa.hibernate.SpringImplicitNamingStrategy</value>
  </property>
  <property>
       <key>hibernate.show_sql</key>
       <value>true</value>
  </property>
```

and Hibernate log the create table sql:

```
 create table user (user_id bigint generated by default as identity, password varchar(255) not null, user_name varchar(255) not null, primary key (user_id))
```

but in h2, the actual tableName is **USER**, and field is also ***uppercase***.

It looks like h2 not create table match the sql. 

Because in h2-1.4.199, **DATABASE_TO_UPPER** default value is true, it make all tableName and field become uppercase. 

so, i make so hack code in **JPADatabase** :

```
 info.put("DATABASE_TO_UPPER", "false");
 connection = new org.h2.Driver().connect("jdbc:h2:tcp://localhost/~/jooq", info);
```
* using remote h2 server to observe what table was generated

This time, uppercase problem is solved. But it create **two** table:

```
CREATE CACHED TABLE "PUBLIC"."User"(
    "userId" BIGINT DEFAULT (NEXT VALUE FOR "PUBLIC"."SYSTEM_SEQUENCE_C411A8A7_925C_4660_B3E8_152DD64E00CC") NOT NULL NULL_TO_DEFAULT SEQUENCE "PUBLIC"."SYSTEM_SEQUENCE_C411A8A7_925C_4660_B3E8_152DD64E00CC",
    "password" VARCHAR(255) NOT NULL,
    "userName" VARCHAR(255) NOT NULL
)

CREATE CACHED TABLE "PUBLIC"."user"(    
"user_id"  BIGINT DEFAULT (NEXT VALUE FOR  "PUBLIC"."SYSTEM_SEQUENCE_D3E8A532_0D26_4E59_A174_D3B719714ED0") NOT  NULL NULL_TO_DEFAULT SEQUENCE  "PUBLIC"."SYSTEM_SEQUENCE_D3E8A532_0D26_4E59_A174_D3B719714ED0",   
 "password" VARCHAR(255) NOT NULL,   
 "user_name" VARCHAR(255) NOT NULL
)

```

But the sql output by hibernate has not change, only **user** create sql has be log. I don't know
who the bug is.
Finally, generate class reference the **User** table, not **user**.
This problem maybe abount entity definition:

```
@Getter
@Setter
@ToString
@Entity
@Table
public class User {

    @Id
    @Column
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long userId;

    @Column(nullable = false)
    private String userName;

    @Column(nullable = false)
    private String password;
}
```

I am not explicitly specified name in @Table and @Column.
After specified name in entity:

```
@Getter
@Setter
@ToString
@Entity
@Table(name = "user")
public class User {

    @Id
    @Column(name = "user_id")
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long userId;

    @Column(name = "user_name", nullable = false)
    private String userName;

    @Column(name = "password", nullable = false)
    private String password;
}
```

The table create in h2 become **user**, and generate class is expected.

**Comment by knutwannheden** [comment]
The fact that a statement like `create table user ...` results in a table with name `USER` is quite normal for databases to do, since the name was not quoted.

I also looked into this and am a bit surprised by what Hibernate is doing. I have not debugged it very thoroughly and don't yet quite understand how it exactly derives the table names, but I found one more property while investigating this. Maybe you could check if it helps in your project if you add this to your pom.xml:

```xml
<property>
    <key>hibernate.globally_quoted_identifiers</key>
    <value>true</value>
</property>
```

**Comment by knutwannheden** [comment]
The thing I don't understand is why `SpringPhysicalNamingStrategy#isCaseInsensitive()` has been hard-coded to return `false`. This will always result in lower-case identifiers being generated by Hibernate. Regardless of what you put in the `@Table` annotation's `name` property. So this strategy isn't really compatible with the `hibernate.globally_quoted_identifiers` property I mentioned.

Thus, if the names in your JPA entities are all correct, I suggest you try with only adding the mentioned `hibernate.globally_quoted_identifiers` property to your pom.xml.

Let me know how it goes.

**Comment by scolia** [comment]
I am try to add **hibernate.globally_quoted_identifiers**, still generated two tables.
but after remove call ***loadAttributeConverters***, generated table become one.
Meybe some bug in ***loadAttributeConverters*** method.

**Comment by knutwannheden** [comment]
> I am try to add hibernate.globally_quoted_identifiers, still generated two tables.
but after remove call loadAttributeConverters, generated table become one.
Meybe some bug in loadAttributeConverters method.

Unfortunately I cannot reproduce this from your description. It would be very helpful if you could report a new issue for this along with an MCVE (https://github.com/jOOQ/jOOQ-mcve). This might allow us to fix this for the 3.12 release.

**Comment by lukaseder** [comment]
I think the original issue (`PUBLIC` schema) has been addressed via several possible workarounds. We'll fix that issue more thoroughly by preventing the generation of this `PUBLIC` schema: https://github.com/jOOQ/jOOQ/issues/9015
